README.md
# Path of Building

Welcome to Path of Building, an offline build planner for Path of Exile!
## Features
* Comprehensive offence + defence calculations:
  * Calculate your skill DPS, damage over time, life/mana/ES totals and much more!
  * Can factor in auras, buffs, charges, curses, monster resistances and more, to estimate your effective DPS
  * Also calculates life/mana reservations
  * Shows a summary of character stats in the side bar, as well as a detailed calculations breakdown tab which can show you how the stats were derived
  * Supports all skills and support gems, and most passives and item modifiers
    * Throughout the program, supported modifiers will show in blue and unsupported ones in red
  * Full support for minions
* Passive skill tree planner:
  * Support for jewels including most radius/conversion jewels
  * Features alternate path tracing (mouse over a sequence of nodes while holding shift, then click to allocate them all)
  * Fully intergrated with the offence/defence calculations; see exactly how each node will affect your character!
  * Can import PathOfExile.com and PoEPlanner.com passive tree links; links shortened with PoEURL.com also work
* Skill planner:
  * Add any number of main or supporting skills to your build
  * Supporting skills (auras, curses, buffs) can be toggled on and off
  * Automatically applies Socketed Gem modifiers from the item a skill is socketed into
  * Automatically applies support gems granted by items
* Item planner:
  * Add items from in game by copying and pasting them straight into the program!
  * Automatically adds quality to non-corrupted items
  * Fully intergrated with the offence/defence calculations; see exactly how much of an upgrade a given item is!
  * Contains a searchable database of all uniques that are currently in game (and some that aren't yet!)
    * You can choose the modifier rolls when you add a unique to your build
    * Includes all league-specific items and legacy variants
  * Features an item crafting system:
    * You can select from any of the game's base item types
	* You can select prefix/suffix modifiers from lists
	* Custom modifiers can be added, with Master and Essence modifiers available
  * Also contains a database of rare item templates:
    * Allows you to create rare items for your build to approximate the gear you will be using
    * Choose which modifiers appear on each item, and the rolls for each modifier, to suit your needs
    * Has templates that should cover the majority of builds (inb4 'why is there no coral amulet?')
* Other features:
  * You can import passive tree, items, and skills from existing characters
  * Share builds with other users by generating a share code
  * Automatic updating; most updates will only take a couple of seconds to apply
  * Somewhat more open source than usual (look in %ProgramData%\Path of Building if you're interested)
  
## Download
Head over to the [Releases](https://github.com/Openarl/PathOfBuilding/releases) page to download the installer.

## Donate
If you'd like to help support the development of Path of Building, I have a [Patreon page](https://www.patreon.com/openarl).

## Screenshots
![ss1](https://cloud.githubusercontent.com/assets/19189971/18089779/f0fe23fa-6f04-11e6-8ed7-ff7d5b9f867a.png)
![ss2](https://cloud.githubusercontent.com/assets/19189971/18089778/f0f923f0-6f04-11e6-89c2-b2c1410d3583.png)
![ss3](https://cloud.githubusercontent.com/assets/19189971/18089780/f0ff234a-6f04-11e6-8c88-6193fe59a5c4.png)


CHANGELOG.md
# Changelog

### 1.4.65 - 2017/12/07
Apologies for the lack of updates recently; I've been very busy. I'll try and manage a few more updates over the
coming weeks, but I can't make any promises yet.
 * Passive tree updated to 3.1
 * You can now rename builds and folders to change only the case of letters
 * Node tooltips now correctly update when cancelling alternate path tracing
 * Fixed Discharge's damage penalty when triggered
 * Fixed Multistrike's attack speed bonus to only apply to Melee attacks
 * Fixed various Skeleton-related modifiers that were being recognised but were not functioning correctly
 * Fixed issue where the program's UI wouldn't be correctly scaled when opened in a non-maximised state

### 1.4.64 - 2017/10/01
 * Trap Throwing Time, Mine Laying Time, and Totem Placement Time are now calculated and shown in the sidebar
    * Special thanks to aggixx for measuring the base time of those animations
 * Trap Cooldown is now shown in the sidebar (in addition to the Calcs tab)
 * Trap Trigger Radius and Mine Detonation Radius are now calculated and shown in the Calcs tab
 * Added support for Vaal Breach
 * All Configuration tab options upon which any Support gems depend are now permanently visible, even if enabling
   them would have no effect
 * Corrected the "Elemental Resistances while on Low Life" stat on Honourhome
 * The Melee Damage buff from Phase Run now correctly excludes Totem skills

This update also reworked the program's window initialisation code.
The most visible change is that the program's main window now opens while the program is initialising, but this
rework is primarily intended to solve three uncommon issues:
 * The program would crash when launched on a non-primary monitor on certain systems
 * The program's UI would be offset when running on systems with certain Intel HD Graphics driver versions
 * The program would crash when launched using Wine

### 1.4.63 - 2017/09/16
 * Added descriptions for support gems
 * The Caustic Cloud from Beacon of Corruption is now correctly affected by Area Damage modifiers on the minion
 * Gaining immunity to Curses now correctly prevents self-Curses from applying
 * Buffs granted by support gems are no longer incorrectly affected by buff effect modifiers of the linked skill
 * Fixed issue causing gem sorting and stat differences to be incorrect when Empower/Enhance/Enlighten are selected

### 1.4.62 - 2017/09/01
 * The instant Leech modifier on Atziri's Acuity should now be recognised correctly
 * Fixed issue preventing modifiers to the damage of Channelling skills from applying to Damage over Time
 * Fixed issue causing the Innervation buff to apply regardless of the setting in the Configuration tab

### 1.4.61 - 2017/08/21
 * The Secondary Durations for Blight, Frost Bomb and Phase Run are now calculated and shown in the Calcs tab
 * Added an option to the Configuration tab for "# of Shocked Enemies Killed Recently"
 * Added support for the "Your Spells are disabled" modifier on Gruthkul's Pelt
 * Poison/Bleed Chance on weapons is now correctly local

### 1.4.60 - 2017/08/21
 * The main Socket Group selector in the sidebar now shows the Socket Group tooltip when you hover over it
 * Updated the skill data for Charged Dash to reflect the changes made in 3.0.1
 * Modifiers that apply when holding a Shield now correctly apply when Necromantic Aegis is allocated
 * The stat comparison for Total DPS inc. Poison is now more intuitive when gaining or losing the ability to Poison
 * Updated the "Is the enemy a Boss?" option to remove the Ailment Duration modifiers from Shaper/Guardian
 * Corrected the base Energy Shield roll on Martyr's Crown
 * Corrected the Critical Strike Multiplier penalty on Ungil's Harmony
 * Updated the Poison Chance modifiers on Snakebite and Cospri's Will

### 1.4.59 - 2017/08/14
With this update, new builds will default to 3.0, and the version selection dialog will no longer display.
Builds can still be converted to 2.6 via the Configuration tab. All 2.6 builds will continue to work as normal, 
however from this point some new features may only be available for 3.0 builds.
 * The Helmet enchantments for the new skills are now available in the item enchanting system
 * The resistance penalties from completing Act 5/10 can now be disabled using a new option in the Configuration tab
 * Removed the attack rate cap for Blink/Mirror Arrow clones, which is no longer present in 3.0
 * An explanatory message is now shown in the sidebar if the main skill is disabled (e.g if no compatible weapon is equipped)
 * Fixed the Burning Damage roll on Pyre
 * Fixed the flat Physical Damage rolls on Widowmaker
 * Fixed the Elemental Resistances roll on Immortal Flesh
 * Fixed issue preventing Socketed Gem modifiers from applying to gems socketed into the alternate weapon set

### 1.4.58 - 2017/08/09
 * Added all of the new uniques
 * Added support for the Void Gaze skill granted by Eber's Unification
 * Added support for the Storm Cascade skill granted by The Rippling Thoughts
 * The other skills granted by the new uniques are only partially supported at the moment (no support for the minions)
 * Charged Dash now has a "Travel distance" option in the Configuration tab
 * Updated the total available passive skill points
 * Burn faster/Burn slower should both now be calculated correctly
 * Modifiers to life/mana/ES recovery rate should now only affect recovery over time
 * The build list now uses natural sort order (so "Foo 50" comes before "Foo 100")
 * The gem selection dropdown now accepts "active" as a filter keyword in additional to other gem tags (such as "support")

For 2.6 builds:
 * The link created when exporting the passive tree now opens in the 2.6.2 version of the offical passive tree viewer

### 1.4.57 - 2017/08/05
 * Now that 3.0 is live, the warning that was shown before importing to 3.0 builds is now shown for 2.6 builds instead

Export/skills.lua
loadStatFile("skill_stat_descriptions.txt")

local statMap = { }
do
	local lastStat
	for line in io.lines("Skills/statmap.ini") do
		local statName = line:match("^%[([^]]+)%]$")
		if statName then
			if lastStat and not lastStat.def then
				statMap[statName] = lastStat
			else
				statMap[statName] = { mult = 1 }
				lastStat = statMap[statName]
			end
		else
			local key, value = line:match("^(%a+) = (.+)$")
			if key == "mod" then
				lastStat.def = value
			elseif key == "mult" then
				lastStat.mult = tonumber(value)
			elseif key == "div" then
				lastStat.mult = 1 / tonumber(value)
			end
		end
	end
end

local function addMod(mods, statKey, value)
	local mod = {
		key = statKey, 
		id = Stats[statKey] and Stats[statKey].Id,
		mult = 1,
		val = value,
		levels = { value },
	}
	if mod.id then
		local map = statMap[mod.id]
		if map then
			mod.def = map.def
			mod.mult = map.mult
		end
	else
		mod.def = statKey
	end
	table.insert(mods, mod)
	return mod
end

-- Skill types:
-- 1	Attack
-- 2	Spell
-- 3	Fires projectiles
-- 4	Dual wield skill (Only on Dual Strike)
-- 5	Buff
-- 6	Can use while dual wielding
-- 7	Only uses main hand
-- 8	Combines both weapons? (Only on Cleave)
-- 9	Minion skill
-- 10	Spell with hit damage
-- 11	Area
-- 12	Duration
-- 13	Shield skill
-- 14	Projectile damage
-- 15	Mana cost is reservation
-- 16	Mana cost is percentage
-- 17	Skill can be trap?
-- 18	Spell can be totem?
-- 19	Skill can be mine?
-- 20	Causes status effects (Only on Herald of Ash, allows Elemental Proliferation)
-- 21	Creates minions
-- 22	Attack can be totem?
-- 23	Chaining
-- 24	Melee
-- 25	Single target melee
-- 26	Spell can multicast
-- 27	?? (On auras, searing bond, tempest shield, blade vortex and others)
-- 28	Attack can multistrike
-- 29	Burning
-- 30	Totem
-- 31	?? (On Molten Shell + Glove Thunder, applied by Blasphemy)
-- 32	Curse
-- 33	Fire skill
-- 34	Cold skill
-- 35	Lightning skill
-- 36	Triggerable spell
-- 37	Trap
-- 38	Movement
-- 39	Cast
-- 40	Damage over Time
-- 41	Mine
-- 42	Triggered spell
-- 43	Vaal
-- 44	Aura
-- 45	Lightning spell
-- 46	?? (Not on any skills or supports)
-- 47	Triggered attack
-- 48	Projectile attack
-- 49	Minion spell
-- 50	Chaos skill
-- 51	?? (Not on any skills, excluded by Faster/Slower Projectiles)
-- 52	?? (Only on Contagion, allows Iron Will)
-- 53	?? (Only on Burning Arrow/Vigilant Strike, allows Inc AoE + Conc Effect)
-- 54	Projectile? (Not on any skills, allows all projectile supports)
-- 55	?? (Only on Burning Arrow/VBA, allows Inc/Red Duration + Rapid Decay)
-- 56	Projectile attack? (Not on any skills, allows projectile attack supports)
-- 57	?? Same as 47
-- 58	Channelled
-- 59	?? (Only on Contagion, allows Controlled Destruction)
-- 60	Cold spell
-- 61	Granted triggered skill (Prevents trigger supports, trap, mine, totem)
-- 62	Golem
-- 63	Herald

local weaponClassMap = {
	[6] = "Claw",
	[7] = "Dagger",
	[8] = "Wand",
	[9] = "One Handed Sword",
	[10] = "Thrusting One Handed Sword",
	[11] = "One Handed Axe",
	[12] = "One Handed Mace",
	[13] = "Bow",
	[14] = "Staff",
	[15] = "Two Handed Sword",
	[16] = "Two Handed Axe",
	[17] = "Two Handed Mace",
	[32] = "Sceptre",
	[36] = "None",
}

local directiveTable = { }

-- #noGem
-- Disables the gem component of the next skill
directiveTable.noGem = function(state, args, out)
	state.noGem = true
end

local function buildStatsPerLevel(grantedKey)
	local levelsStats = {}

	local levels = GrantedEffectsPerLevel.GrantedEffectsKey(grantedKey)

	for level, levelKey in ipairs(levels) do
		local effect = GrantedEffectsPerLevel[levelKey]
		local levelStats = {}

		for statNum, statKey in pairs(effect.StatsKeys) do
			local stat = Stats[statKey]
			
			table.insert(levelStats, {
				stat = stat,
				value = effect["Stat" .. statNum .. "Value"],
				float = effect["Stat" .. statNum .. "Float"]
			})
		end

		for _, statKey in pairs(effect.StatsKeys2) do
			local stat = Stats[statKey]
			
			table.insert(levelStats,  {
				stat = stat,
				value = 1,
				float = 1.0
			})
		end

		levelsStats[level] = levelStats
	end

	return levelsStats
end

-- computes an input to #describeStats 
-- we assume that quality bonus doesn't change with levels although
-- the bonus is per level
local function buildQualityStats(grantedKey)
	local stats = {}

	local levels = GrantedEffectsPerLevel.GrantedEffectsKey(grantedKey)
	local effect = GrantedEffectsPerLevel[levels[1]]

	for statNum, statKey in pairs(effect.Quality_StatsKeys) do
		local stat = Stats[statKey]
			-- stat value per 1000 quality to per 1 quality
		local value = effect.Quality_Values[statNum] / 1000

		stats[stat.Id] = {
			-- assume max quality for now to get the correct translation
			-- correct behavior would be to assume 1 quality
			-- and count up until we get a description
			-- e.g. barrage q gives .5% inc proj damage per quality
			-- but it actually only considers 1% inc proj damage per 2 q
			min = value * 20,
			max = value * 20
		}
	end

	return stats
end

-- computes an input to #describeStats from #buildStatsPerLevel
-- basically aggregates over levels and finds for each stat the min and max value
local function aggregateStats(levelsStats)
	local stats = {}

	for level, levelStats in ipairs(levelsStats) do
		for _, levelStat in ipairs(levelStats) do
			local statId = levelStat.stat.Id

			if not stats[statId] then
				stats[statId] = {
					min = math.huge,
					max = -math.huge
				}
			end

			stats[statId] = {
				min = math.min(stats[statId].min, levelStat.value),
				max = math.max(stats[statId].max, levelStat.value)
			}

			if levelStat.value == nil then
				print('nil!')
			end
		end
	end

	return stats
end

-- #skill <GrantedEffectId> [<Display name>]
-- Initialises the gem data and emits the skill header
directiveTable.skill = function(state, args, out)
	local grantedId, displayName = args:match("(%w+) (.+)")
	if not grantedId then
		grantedId = args
		displayName = args
	end
	out:write('skills["', grantedId, '"] = {\n')
	local grantedKey = GrantedEffects.Id(grantedId)[1]
	local granted = GrantedEffects[grantedKey]
	if not granted then
		print('Unknown GE: "'..grantedId..'"')
	end
	local skillGemKey = SkillGems.GrantedEffectsKey(grantedKey)[1]
	local gem = { }
	state.gem = gem
	if skillGemKey and not state.noGem then
		local skillGem = SkillGems[skillGemKey]
		out:write('\tname = "', BaseItemTypes[skillGem.BaseItemTypesKey].Name:gsub(" Support",""), '",\n')
		local tagNames = { }
		out:write('\tgemTags = {\n')
		for i, tagKey in ipairs(skillGem.GemTagsKeys) do
			out:write('\t\t', GemTags[tagKey].Id, ' = true,\n')
			if #GemTags[tagKey].Tag > 0 then
				table.insert(tagNames, GemTags[tagKey].Tag)
			end
		end
		out:write('\t},\n')
		out:write('\tgemTagString = "', table.concat(tagNames, ", "), '",\n')
		out:write('\tgemStr = ', skillGem.Str, ',\n')
		out:write('\tgemDex = ', skillGem.Dex, ',\n')
		out:write('\tgemInt = ', skillGem.Int, ',\n')
	else
		if displayName == args and not granted.IsSupport then
			displayName = ActiveSkills[granted.ActiveSkillsKey].DisplayedName
		end
		out:write('\tname = "', displayName, '",\n')
		out:write('\thidden = true,\n')
	end
	state.noGem = false
	gem.baseFlags = { }
	gem.mods = { }
	local modMap = { }
	gem.levels = { }
	gem.global = "nil"
	gem.curse = "nil"
	out:write('\tcolor = ', granted.Unknown0, ',\n')
	if granted.IsSupport then
		gem.isSupport = true
		out:write('\tsupport = true,\n')
		out:write('\trequireSkillTypes = { ')
		for _, type in ipairs(granted.Data0) do
			out:write(type, ', ')
		end
		out:write('},\n')
		out:write('\taddSkillTypes = { ')
		for _, type in ipairs(granted.Data1) do
			out:write(type, ', ')
		end
		out:write('},\n')
		out:write('\texcludeSkillTypes = { ')
		for _, type in ipairs(granted.Data2) do
			out:write(type, ', ')
		end
		out:write('},\n')
	else
		local activeSkill = ActiveSkills[granted.ActiveSkillsKey]
		if #activeSkill.Description > 0 then
			out:write('\tdescription = "', activeSkill.Description, '",\n')
		end
		out:write('\tskillTypes = { ')
		for _, type in ipairs(activeSkill.ActiveSkillTypes) do
			out:write('[', type, '] = true, ')
		end
		out:write('},\n')
		if activeSkill.Unknown19[1] then
			out:write('\tminionSkillTypes = { ')
			for _, type in ipairs(activeSkill.Unknown19) do
				out:write('[', type, '] = true, ')
			end
			out:write('},\n')
		end
		local weaponTypes = { }
		for _, classKey in ipairs(activeSkill.WeaponRestriction_ItemClassesKeys) do
			if weaponClassMap[classKey] then
				weaponTypes[weaponClassMap[classKey]] = true
			end
		end
		if next(weaponTypes) then
			out:write('\tweaponTypes = {\n')
			for type in pairs(weaponTypes) do
				out:write('\t\t["', type, '"] = true,\n')
			end
			out:write('\t},\n')
		end
		if activeSkill.SkillTotemId ~= 16 then
			out:write('\tskillTotemId = ', activeSkill.SkillTotemId, ',\n')
		end
		local typeFlag = { }
		for _, type in ipairs(activeSkill.ActiveSkillTypes) do
			typeFlag[type] = true
		end
		if typeFlag[32] then
			gem.global = '{ type = "GlobalEffect", effectType = "Curse" }'
			gem.curse = gem.global
		elseif typeFlag[44] then
			gem.global = '{ type = "GlobalEffect", effectType = "Aura" }'
		elseif typeFlag[5] or typeFlag[31] then
			gem.global = '{ type = "GlobalEffect", effectType = "Buff" }'
		end
		addMod(gem.mods, 'skill("castTime", {val})', granted.CastTime / 1000)
	end
	for _, key in ipairs(GrantedEffectsPerLevel.GrantedEffectsKey(grantedKey)) do
		local level = { }
		local levelRow = GrantedEffectsPerLevel[key]
		level.level = levelRow.Level
		table.insert(gem.levels, level)
		local function addLevelMod(statKey, value, forcePerLevel)
			local mod = gem.mods[modMap[statKey]]
			if mod then
				if value ~= mod.val then
					mod.perLevel = true
				end
			else
				modMap[statKey] = #gem.mods + 1
				addMod(gem.mods, statKey)
				mod = gem.mods[modMap[statKey]]
				mod.val = value
			end
			if forcePerLevel then
				mod.perLevel = true
			end
			mod.levels[levelRow.Level] = value
		end
		if not granted.IsSupport then
			addLevelMod('skill("levelRequirement", {val})', levelRow.LevelRequirement, true)
		else
			addLevelMod("nil", levelRow.LevelRequirement, true)
		end
		if levelRow.ManaCost and levelRow.ManaCost ~= 0 then
			addLevelMod('skill("manaCost", {val})', levelRow.ManaCost)
		end
		if levelRow.ManaMultiplier ~= 100 then
			addLevelMod('mod("ManaCost", "MORE", {val})', levelRow.ManaMultiplier - 100)
		end
		if levelRow.DamageEffectiveness ~= 0 then
			addLevelMod('skill("damageEffectiveness", {val})', levelRow.DamageEffectiveness / 100 + 1)
		end
		if levelRow.CriticalStrikeChance ~= 0 then
			addLevelMod('skill("CritChance", {val})', levelRow.CriticalStrikeChance / 100)
		end
		if levelRow.DamageMultiplier and levelRow.DamageMultiplier ~= 0 then
			addLevelMod('skill("baseMultiplier", {val})', levelRow.DamageMultiplier / 10000 + 1)
		end
		if levelRow.ManaReservationOverride ~= 0 then
			addLevelMod('skill("manaCostOverride", {val})', levelRow.ManaReservationOverride)
		end
		if levelRow.Cooldown and levelRow.Cooldown ~= 0 then
			addLevelMod('skill("cooldown", {val})', levelRow.Cooldown / 1000)
		end
		for i, statKey in ipairs(levelRow.StatsKeys) do
			addLevelMod(statKey, levelRow["Stat"..i.."Value"])
		end
		for i, statKey in ipairs(levelRow.StatsKeys2) do
			addLevelMod(statKey)
		end
		if not gem.qualityMods then
			gem.qualityMods = { }
			for i, statKey in ipairs(levelRow.Quality_StatsKeys) do
				addMod(gem.qualityMods, statKey, levelRow.Quality_Values[i] / 1000)
			end
		end
	end

	local levelsStats = buildStatsPerLevel(grantedKey)
	local levelStatsAggregated = aggregateStats(levelsStats)
	local effectsDescription = describeStats(levelStatsAggregated)
	
	out:write('\teffects = {\n')
	for i, effectDescription in pairs(effectsDescription) do 
		out:write('\t\t[' .. i .. '] = "' .. effectDescription .. '",\n')
	end
	out:write('\t},\n')

	local qualityStats = buildQualityStats(grantedKey)
	local qualityEffectsDescription = describeStats(qualityStats)
	out:write('\teffectsQuality = {\n')
	for i, effectDescription in pairs(qualityEffectsDescription) do 
		out:write('\t\t[' .. i .. '] = "' .. effectDescription .. '",\n')
	end
	out:write('\t},\n')

	if grantedId == 'Barrage' then
		for stat, range in pairs(qualityStats) do
			print(stat, range.min, range.max)
		end
	end
end

-- #global <Buff|Aura|Debuff|Curse>
-- Sets the global effect tag for this skill
directiveTable.global = function(state, args, out)
	local gem = state.gem
	gem.global = '{ type = "GlobalEffect", effectType = "'..args..'" }'
end

-- #flags <flag>[ <flag>[...]]
-- Sets the base flags for this active skill
directiveTable.flags = function(state, args, out)
	local gem = state.gem
	for flag in args:gmatch("%a+") do
		table.insert(gem.baseFlags, flag)
	end
end

-- #baseMod <mod definition>
-- Adds a base modifier to the skill
directiveTable.baseMod = function(state, args, out)
	local gem = state.gem
	addMod(gem.mods, args)
end

-- #levelMod <mod definition>==<val>[ <val>[...]]
-- Adds a per-level modifier to the skill
directiveTable.levelMod = function(state, args, out)
	local gem = state.gem
	local def, vals = args:match("(.*)==(.*)")
	local mod = addMod(gem.mods, def)
	mod.perLevel = true
	local i = 1
	for _, level in ipairs(gem.levels) do
		local s, e, val = vals:find("([%+%-]?[%d%.]+)", i)
		mod.levels[level.level] = tonumber(val)
		i = e + 1
	end
end

-- #setMod <StatId>==<mod definition[;<mult|div>=<val>]
-- Sets or overrides the mapping of the given stat
directiveTable.setMod = function(state, args, out)
	local gem = state.gem
	local id, def = args:match("(.*)==(.*)")
	for _, list in ipairs({gem.mods,gem.qualityMods}) do
		for _, mod in ipairs(list) do
			if mod.id == id then
				local name, mult = def:match("(.*);mult=(.*)")
				if name then
					mod.def = name
					mod.mult = tonumber(mult)
				else
					local name, div = def:match("(.*);div=(.*)")
					if name then
						mod.def = name
						mod.mult = 1 / tonumber(div)
					else
						mod.def = def
					end
				end
			end
		end
	end
end

-- #mods
-- Emits the skill modifiers
directiveTable.mods = function(state, args, out)
	local gem = state.gem
	if not gem.isSupport then
		out:write('\tbaseFlags = {\n')
		for _, flag in ipairs(gem.baseFlags) do
			out:write('\t\t', flag, ' = true,\n')
		end		
		out:write('\t},\n')
	end
	out:write('\tbaseMods = {\n')
	for _, mod in ipairs(gem.mods) do
		if not mod.perLevel then
			out:write('\t\t')
			if mod.def and mod.def ~= "nil" then
				out:write(mod.def:gsub("{val}",(mod.val or 0)*mod.mult):gsub("{global}",gem.global):gsub("{curse}",gem.curse), ', ')
			end
			if mod.id then
				out:write('--"', mod.id, '" = ', (mod.val or "?"))
			end
			out:write('\n')
		end
	end
	out:write('\t},\n')
	out:write('\tqualityMods = {\n')
	for _, mod in ipairs(gem.qualityMods) do
		out:write('\t\t')
		if mod.def then
			out:write(mod.def:gsub("{val}",mod.levels[1]*mod.mult):gsub("{global}",gem.global):gsub("{curse}",gem.curse), ', ')
		end
		if mod.id then
			out:write('--"', mod.id, '" = ', mod.levels[1])
		end
		out:write('\n')
	end
	out:write('\t},\n')
	out:write('\tlevelMods = {\n')
	local lcol = 1
	for _, mod in ipairs(gem.mods) do
		if mod.perLevel then
			out:write('\t\t')
			if mod.def then
				out:write('[', lcol, '] = ', mod.def:gsub("{val}","nil"):gsub("{global}",gem.global):gsub("{curse}",gem.curse), ', ')
				if mod.id then
					out:write('--"', mod.id, '"')
				end
				out:write('\n')
			else
				out:write('--[', lcol, '] = "', mod.id, '"\n')
			end
			mod.col = lcol
			lcol = lcol + 1
		end
	end
	out:write('\t},\n')
	out:write('\tlevels = {\n')
	for _, level in ipairs(gem.levels) do
		out:write('\t\t[', level.level, '] = { ')
		for _, mod in ipairs(gem.mods) do
			if mod.perLevel then
				if mod.levels[level.level] then
					out:write(tostring(mod.levels[level.level] * mod.mult), ', ')
				else
					out:write('nil, ')
				end
			end
		end
		out:write('},\n')
	end
	out:write('\t},\n')
	out:write('}')
	state.gem = nil
end

for _, name in pairs({"act_str","act_dex","act_int","other","glove","minion","spectre","sup_str","sup_dex","sup_int"}) do
	processTemplateFile("Skills/"..name, directiveTable)
end

os.execute("xcopy Skills\\act_*.lua ..\\Data\\3_0\\Skills\\ /Y /Q")
os.execute("xcopy Skills\\sup_*.lua ..\\Data\\3_0\\Skills\\ /Y /Q")
os.execute("xcopy Skills\\other.lua ..\\Data\\3_0\\Skills\\ /Y /Q")
os.execute("xcopy Skills\\glove.lua ..\\Data\\3_0\\Skills\\ /Y /Q")

print("Skill data exported.")